### ISSUE TITLE
Create functions to check data from Wikipedia

### ISSUE THREAD

**Issue by MansMeg** [MansMeg/SwedishPolls#64]
https://en.wikipedia.org/wiki/Swedish_general_election,_2018

**Comment by MansMeg** [comment]
Done by @tommiehansen.

Maybe you could add the code you use to the repository in a subfolder?

**Comment by tommiehansen** [comment]
Maybe. The current array_merge() function being used is slighly ugly since it causes some obvious sorting errors in the data. I think i may go back to just overwriting everything using the other sort order.

Nothing says that one could have alternatives to Polls.csv though (ie an alternative that strictly only uses *real* available data) etc. You could then use that data for Polls.csv as you, or others, see fit. It would be much easier to use that data as a template for new updates to Polls.csv or even use this new data if one doesn't care as much about publishing date.

I'll work on it some more and clean up the code so that it is more ready for commits.

**Comment by MansMeg** [comment]
Great!

**Comment by tommiehansen** [comment]
Have been adding stuff; it's mostly just a rewrite/cleanup of the already working code and some actual folder structure(s) etc:
https://github.com/tommiehansen/SwedishPolls

The additions only reside within the folder `/SwedishPolls/External/`

Currently only contains the parsed data from Wikipedia @ https://en.wikipedia.org/wiki/Opinion_polling_for_the_Swedish_general_election,_2018

But once one have all the parsing logic (and fixes...) for that the rest is pretty simple.

Any idea to also add data from https://sv.wikipedia.org/wiki/Opinionsm%C3%A4tningar_inf%C3%B6r_riksdagsvalet_i_Sverige_2018
...?

A problem with that data is that its missing the more useful `collectPeriodFrom` and `collectPeriodTo`. Those two values is imo more interesting then `PublDate` though since obvious reasons. :)

**Comment by MansMeg** [comment]
Great! Would it be possible to have an example script file you could try to run to parse this data? So it would be possible to just run something like:
php --script_name.php --"wikpedia.url/polls.url.address.html"
That would make it easier for others to use the script. But Im not very familiar with PHP.

I think it would be great to add data from the other poll as well?

**Comment by tommiehansen** [comment]
The data already gets parsed? The result resides within `External/Data/`.
If one want to parse the parsed data just load the .csv or .sqlite and do stuff? It's as hard/easy as parsing any other data format. :-)

Since the data at Wikipedia isn't really standardized one can't have dynamic script file(s) like the example you give but rather need to be build a specific parser per specific page.  Just getting the page
`wikipedia.url/random.url.html` isnt' really a thing since the thing is to parse the data and make something useful with it (in this case standardizing certain things and make it more compatible with Polls.csv even if not 100%).

Therefor the runners `PHP/polls_wikipeda.php` and `PHP/wikipedia.php` is more specific and thus isn't compatible with other sources that has other data and thus need other parsing.

Just compare:
https://en.wikipedia.org/wiki/Opinion_polling_for_the_Swedish_general_election,_2018
...and...
https://sv.wikipedia.org/wiki/Opinionsm%C3%A4tningar_inf%C3%B6r_riksdagsvalet_i_Sverige_2018

Note how the standard differs. Not only the splitting of 1x table/company but also the data itself that is missing the more important collectPeriod. ie "Jul 2017" and thus really isn't that usable. The first one is usable though.

I haven't found other data sources that could be useful, but i'm looking into it. I'm thinking of directly scraping the sources but this might be a bit too tedious since all the original sources differs quite a lot.

**Comment by MansMeg** [comment]
Alright, but I was more thinking of parsing new data. Say in two years, how would I use the php code to parse the data from Wikipedia (and convert it to the csv file)? I.e. how would I use the code (as simply as possible).

**Comment by tommiehansen** [comment]
Everything in the root of `External/PHP` are fetchers/parsers?

See the README.md under:
https://github.com/tommiehansen/SwedishPolls/tree/master/External/PHP

Your question seem to be "How do i run PHP"? Since PHP is the most popular dynamic language on the web i don't see much reason to write how to run PHP. The simplest way to run the stuff is to just install apache and php via Bash (linux/win 10/osx).

To run just cd into PHP-folder and do ie `php wikipedia.php`.

**Comment by MansMeg** [comment]
Ok. I'll try it. 

PHP may be a popular dynamic language, but I have never used it or programmed in it. =) I

**Comment by tommiehansen** [comment]
Yeah, ok. Just google for `bash apache php 7` and if under win10 google `bash apache php 7 windows 10`. It's pretty straight forward to get up and running.

**Comment by tommiehansen** [comment]
I'm reworking how the 'merge' -thing works to better standardize things. The other code got a bit bloated and there is better way to do things etc.

One problem that will exist though is that the merged data will contain duplicates. This due to the fact that sometimes Polls.csv actually has data that is wrong and since the only reliable way to check for dupes is by using 'Year + Company + M, L, C, KD, S' anything where Polls.csv differs for some party will be considered a NON-duplicate. Format Y-m-d cannot be used for checking duplicates since that data differs by a day or two sometimes.

The reason that Year + Month + M, L, C, KD, S, V is used is due to:
- `collectPeriodFrom` and `collectPeriodTo` differ quite a lot between Polls.csv and Wikipedia meaning if one uses that a lot of data would be considered non-duplicate (even if it obviously is).
- We can be quite certain that the Year and Month for `collectPeriodTo` does not differ
- Numbers for parties should not differ else either Polls.csv has it wrong or Wikipedia has it wrong

A current example is Demoskop @ publ. april 2017 where Polls.csv indicate that `S` has the value of `26.6` but the real value is `26.2` according to:
A: http://www.demoskop.se/publicerat/valjarbarometern-april-2/
B: Wikipedia data

It becomes a duplicate since there is no certain way to, programmatically, know what source is correct and what source has it wrong.

**Comment by MansMeg** [comment]
So, a duplicate is not a problem. I see this more as a way to double check the `Polls.csv` against the Wikipedia page. If there are duplicates, we can manually check those and get the correct result. 

I do not think we should use this for automatically updating due to the problems you identified.

**Comment by MansMeg** [comment]
The error in the Demoskop poll has now been fixed.

**Comment by tommiehansen** [comment]
Yes and great. :)

The current merging and conversion got slightly more usable during the night. Now they take a a bounch of arguments meaning that one can convert Polls (or a merge.. or Wikipedia) to csv, tsv and json and specify things such as max to merge `maxmerge` from Wikipedia but also can choose file names during runtime etc.

Here's how that work using the `automaton.sh` bash script:
```
#!/bin/bash

clear

# get core Wikipedia data and Polls.csv data
php polls.php
php wikipedia.php
php sqlite_format.php file=Polls.sqlite
php sqlite_format.php file=Wikipedia.sqlite

# standard merge + create csv
php merge.php
php sqlite_format.php file=Merged.sqlite format=csv file_out=Merged.csv

# create file with only X latest
php merge.php name=Merged_last10.sqlite strict=half-strict maxmerge=10
php sqlite_format.php file=Merged_last10.sqlite format=csv file_out=Merged.last10.csv
```

The limited versions with the `half-strict` mode will be more usable since it'll be more reliable. It's basically more for just performing updates and answering/acting upon "Is there new data?" while the default tries to merge all existing data.

The wikipedia.php also got slightly more complex since it now also fetches data from:
https://en.wikipedia.org/wiki/Opinion_polling_for_the_Swedish_general_election,_2014

**Comment by MansMeg** [comment]
Great! Ill try it asap.

**Comment by tommiehansen** [comment]
Please let me know if you need any help with anything PHP @ cmdline.
What host OS are you running?

PHP should be able to run everything without needing an external web server like Apache if you got the required extensions installed (ie CURL and stuff for PHP DOM).

**Comment by MansMeg** [comment]
Sure. That should work. I run on Mac Os.

**Comment by tommiehansen** [comment]
Oh -- ok. Got it running?
-
I just updated the scripts with more terminal friendly status commands, better handling of arguments for merge.php etc. Now it's mostly cleanup and maybe write some examples or something.

**Comment by MansMeg** [comment]
Haven't tried yet. I'm currently waging a war with CUDA. =)

But if you have a simple example/bash script that would be great. Even better if you could try to get it up and working on Travis.

**Comment by tommiehansen** [comment]
The simple bash script you're referring to already exists as `automate.sh` under `External/PHP/`
Setting up a scheduler to perform A, B or C is something different though.

Travis is for builds, there are no 'builds' @ /External per se since the source code isn't intended to really change at all. It's just a bounch of generators and parsers so there is little to really _check_. The files that get update simply will get updated.

**Comment by MansMeg** [comment]
Cool. Yes, I know. But I use Travis in a wider way. Such as running checks with new data. The code does not change, but the data does, and it is much easier to get this done automatically than manually checking all contributors data.

**Comment by tommiehansen** [comment]
Yeah, i figured since the data isn't added programatically and thus needs checking since any data added could be in any shape or form. Especially since a 0.15% difference is hard to spot but could result in errors etc.

In the case of the scripts there isn't really anything to test though since the stuff that can fail already has checks internally where relevant e.g. (pseudocode):

```
foreach( $wikiData as $array ){
  if( !isset( $array['something'] ) ) {
    // something was not set, we need to do something else
  }
}
```

Because obviously checking and testing needs to be done. Basically **80-90%** of the code consists of testing and checking. Without it everything would be completely useless. 😄

The smallest, and most trivial, parts are the stuff that gets and writes the data. The output is thus a result of all testing and checking.

However i'm writing a sort of 'report' function for the generated SQLite databases. This since some things can be 'odd' but cannot be tested properly due to the fact that there is no visible **erroneous pattern**. Sort of "suspect and may be bad, but maybe not". Mostly this will be a good addition to quickly check/compare the original Polls.csv data and find entries that may be wrong (or find entries in the Wikipedia data that is wrong) etc. It will also be helpful to spot any pattern that could be programmatically implemented in the dupe-checking process.

**Comment by tommiehansen** [comment]
So currently i'm just running this (automaton.sh) from time to time. Haven't really gotten around to create a scheduler for it yet since i'm debating, with myself, how often it should run and if there should be some sort of black-out period (no point in wasting resources by running it between 00:00-10:00 etc).

As you can see there's currently 3x polls missing from `Polls.csv` that the automaton.sh has added to its different output formats.

The `odd_checker` also now saves reports to External/PHP/cache/ as plain HTML-documents (with everything inline eg css/js etc). What it does is basically list things that is "odd" or "might be bad" as i wrote about previously but it does it with 3 sort of checks:

* Check for rows with `NULL` (NA) `PublYearMonth` ( e.g. all from Wikipedia that doesn't have such a value and where Polls.csv hasn't been updated or where Polls.csv simply has missing data)
* +/- 2 rows from above (to manually check if there's a row close by that is wrong etc)
* Rows where next row is the same company and the collectPeriodTo is in same month. Results in a lot of rows containing the month of September (obviously) but also some rows that where one can wonder what data is correct, example screen:
![image](https://user-images.githubusercontent.com/8710941/29360790-dceb9bb0-8284-11e7-8d3a-81445f15f739.png)
_Numbers from Demoskop are odd to say the least. Rows that doesn't contain `PublYearMonth` is the data from Wikipedia_

One can actually read the reports directly via RAWGit, here's some links:
https://rawgit.com/tommiehansen/SwedishPolls/master/External/PHP/cache/Merged.sqlite.report.html
https://rawgit.com/tommiehansen/SwedishPolls/master/External/PHP/cache/Merged_last10.sqlite.report.html
https://rawgit.com/tommiehansen/SwedishPolls/master/External/PHP/cache/Polls.sqlite.report.html

**Comment by MansMeg** [comment]
Cool! So all those polls in https://rawgit.com/tommiehansen/SwedishPolls/master/External/PHP/cache/Polls.sqlite.report.html are "missing" in Polls.csv?

**Comment by tommiehansen** [comment]
No since `Polls.sqlite.report.html` is a report for Polls.csv, wrong link?
However this:
https://rawgit.com/tommiehansen/SwedishPolls/master/External/PHP/cache/Merged.sqlite.report.html
Has rows that's either missing from Polls.csv or that cannot be marked as 'duplicate' for some reason. I just updated the reports to add some clarity to what the different tables mean.

**Comment by MansMeg** [comment]
Great! It is nice to find a couple of old missing polls as well!

**Comment by MansMeg** [comment]
Are you writing this to the Polls files and then doing a PR?

**Comment by tommiehansen** [comment]
Nothing is written to the original Polls.csv -file under `/Data/`. However an alternate version is created under `/External/Data/`. That version is basically Polls.csv with the only difference beeing that it is sorted using `collectPeriodTo`. It's mostly used for merging and to output Polls.csv to different formats other then CSV.

Cannot write the data to original Polls.csv since some values simply do not exist in the Wiki-data.
Overwriting it would also cause a huge amount of "diff" due to sorting. This since Polls.csv uses the field `PublDate` for sorting but this value does not exist for the Wiki-data and is always unknown. Hence it cannot be used for any sorting or any sort-and-merge attempt.

There are a lot of cases like the below example (pseudo):
```
company   publDate        collectPeriodTo
YouGov    2010-01-10      2010-01-05
Sifo      2010-01-09      2010-01-08
```

If sorted by collectPeriodTo Sifo would obviously be #1 and YouGov #2.  If applied to original Polls.csv we would basically establish a new sort order and any newly added rows could be subject to change (basically any manually added rows would have to respect the collectPeriodTo value and thus sometimes add new data on the 3rd row instead of the 2nd row).

tl;dr
It's easier to just use the `/External/` stuff for manually updating Polls.csv like you said before.
Alternatively someone could use `Merged_last10` -files since that one very seldomly will contain erroneous data _(and if one doesn't use published date of any sort for doing something that is...)_

Not ready for PR yet. I think there still are some small niggles to refine thus it would be helpful if you could take some time to check things.

**Comment by tommiehansen** [comment]
Polls.csv is falling behind quite a bit btw.

Currently there are 6 more recent polls missing from Polls.csv as one simply can see by checking https://github.com/tommiehansen/SwedishPolls/blob/82f5e539643f4e7f7caa58c7edfea9dab90f4415/External/Data/Merged.csv

I also saw that @hjnilsson had added most of these.

**Comment by MansMeg** [comment]
Yes. Now it should be up to date.
